src/lib/es5.d.ts

```diff
@@ -180,6 +180,18 @@ interface ObjectConstructor {
       * Prevents the modification of existing property attributes and values, and prevents the addition of new properties.
       * @param o Object on which to lock the attributes.
       */
+    freeze<T>(a: T[]): ReadonlyArray<T>;
+
+    /**
+      * Prevents the modification of existing property attributes and values, and prevents the addition of new properties.
+      * @param o Object on which to lock the attributes.
+      */
+    freeze<T>(f: (...args: any[]) => T): (...args: any[]) => T;

```

Comment:
However, even this definition doesn't work with generic functions and classes. Readonly interface doesn't work well because of some limitations of the current type system. I hope TypeScript will resolve these limitations.

related #9366